### ISSUE TITLE
Intepret Resource into F[A, F[Unit]]

### ISSUE THREAD

**Issue by rossabaker** [typelevel/cats-effect#378]
I had a need to [separate acquire from release](https://github.com/http4s/http4s/blob/18306d575fb7e7363300d514d89305a37107c060/client/src/main/scala/org/http4s/client/DefaultClient.scala#L65-L98) on `Resource`, similar to the `allocate` value that was removed in #278.  Would adding `allocate` back to `Resource` with this interpreter be of any interest?

**Comment by alexandru** [comment]
The problem and why I removed it is that I don't think it's possible.

**Comment by SystemFw** [comment]
Quick tl;dr for the use case: this would allow you to "transfer" finalisers to a different context. 
For example what we had to do in http4s was implementing a proxy server via a client, which needs to take the finaliser that closes the connection after a client call, and moves it to the server

**Comment by SystemFw** [comment]
@alexandru Could you take a look at the interpreter we have and see if there are any glaring issues?

**Comment by SystemFw** [comment]
```scala
  def toHttpApp: HttpApp[F] = Kleisli { req =>
    /* Derived from https://github.com/typelevel/cats-effect/blob/c62c6c76b3066c500fca2f9e0897605bc12eb2a0/core/shared/src/main/scala/cats/effect/Resource.scala */

    // Indirection for calling `loop` needed because `loop` must be @tailrec
    def continue(
        current: Resource[F, Any],
        stack: List[Any => Resource[F, Any]],
        release: F[Unit]): F[(Any, F[Unit])] =
      loop(current, stack, release)

    // Interpreter that knows how to evaluate a Resource data structure;
    // Maintains its own stack for dealing with Bind chains
    @tailrec def loop(
        current: Resource[F, Any],
        stack: List[Any => Resource[F, Any]],
        release: F[Unit]): F[(Any, F[Unit])] =
      current match {
        case Resource.Allocate(resource) =>
          F.bracketCase(resource) {
            case (a, rel) =>
              stack match {
                case Nil => F.pure(a -> rel(ExitCase.Completed).guarantee(release))
                case f0 :: xs => continue(f0(a), xs, rel(ExitCase.Completed).guarantee(release))
              }
          } {
            case (_, ExitCase.Completed) =>
              F.unit
            case ((_, release), ec) =>
              release(ec)
          }
        case Resource.Bind(source, f0) =>
          loop(source, f0.asInstanceOf[Any => Resource[F, Any]] :: stack, release)
        case Resource.Suspend(resource) =>
          resource.flatMap(continue(_, stack, release))
      }

    loop(run(req).asInstanceOf[Resource[F, Any]], Nil, F.unit).map {
      case (a, release) =>
        val resp = a.asInstanceOf[Response[F]]
        resp.copy(body = resp.body.onFinalize(release))
    }
  }
```

**Comment by alexandru** [comment]
I don't know if there's anything wrong with it, we probably have to play with it. And I don't have time this week for anything extra.

Would be cool to add it back, I remember having issues with it though. But if you can do it, of course. We have to ensure correctness.

**Comment by SystemFw** [comment]
Ok, if there's potential interest we can strive to add more test cases (which won't hurt http4s either) I think?

**Comment by rossabaker** [comment]
http4s tests that:
- Release works in the right order on the happy path
- Nested resources are released correctly if an inner acquire fails

Missing test cases:
- Nested resources are released under cancellation.
- ???

**Comment by alexandru** [comment]
One thing to consider is that the reason for why I changed `Resource` is that it wasn't as generic as its restrictions (i.e. `Bracket` and `Sync`), the original implementation assuming `IO`'s behavior for `flatMap`.

So that's kind of the requirement ... if the restriction is `Bracket` or `Sync`, it has to work with streams. I can play with your implementation and `Iterant` if you'd like, `Iterant` implementing `Async` at the moment.

Do a PR and we'll see.

**Comment by SystemFw** [comment]
yeah I think it's most likely to come down to cancelation scenarios, stressing things a lot to find potential weird interleavings (I use property based testing for that).

**Comment by alexandru** [comment]
> if the restriction is Bracket or Sync, it has to work with streams

Not an absolute requirement btw, but if it doesn't work with streams, then we are introducing assumption about behavior that aren't set in stone by laws and that's not cool.

**Comment by SystemFw** [comment]
@rossabaker btw, I'm not entirely clear if your proposal is to change the Resource encoding, or keep it as is, but adding a `toAllocate` interpreter (which is what we're doing now).

@alexandru Note that the interpreter above works with the current resource encoding, we didn't fork it or change it, just added a new op

**Comment by alexandru** [comment]
Do you think there's anything wrong with the current encoding?

It's a public encoding, left it public so we can build interpreters for it and we can't really change it during 1.x.

**Comment by SystemFw** [comment]
> Do you think there's anything wrong with the current encoding?

No, that's what I'm saying. I like the current encoding and I think that as long as `Bind` remains a data type, it can be interpreted with `Stream/Iterant`.

Are you saying that you would like the `toAllocate` interpreter to _also_ work with `Stream/Iterant`?

**Comment by rossabaker** [comment]
I think the current `Resource` encoding is still good and necessary, and even if it's not, needs to stay for binary compatibility.  This is just a way of safely (we hope) interpreting it back to the original `allocate`.

Testing on different effects, like streams, is a very good point.  The http4s use will be `IO` or `Task` unless someone is doing something incredibly exotic, but we need to handle all the effects with a valid instance.

**Comment by SystemFw** [comment]
Ok, I think we're all on the same page, thanks for clarifying :)

**Comment by alexandru** [comment]
Right, so ...

1. `Allocate` is the old encoding
2. `FlatMap` and `Suspend` are there in order to unpack `flatMap` calls

We need to unpack flatMap calls if we want to convert a `Resource` into an equivalent data type like a `Sync` stream.

> Are you saying that you would like the toAllocate interpreter to also work with Stream/Iterant?

Would be nice. Or otherwise we need to consider what restrictions we need to add in order to exclude streams statically.

**Comment by SystemFw** [comment]
> Would be nice. Or otherwise we need to consider what restrictions we need to add in order to exclude streams statically.

Ok, makes sense. Note that `fs2.Stream` is not `Bracket` since `Stream.bracket` is different and more powerful than `F.bracket`, so it is excluded statically, but I definitely don't want to impose that restriction to other streaming implementation. I agree that we should strive to make the `toAllocate` interpreter (better name desired) work for `Iterant`, if at all possible.

Sorry for introducing confusion btw, I don't and never wanted to change the ADT encoding of Resource :)

**Comment by alexandru** [comment]
Can we move this issue forward?

I just noticed that the documentation on `Resource` is currently incorrect, since it lists `allocate`: https://typelevel.org/cats-effect/datatypes/resource.html

So we could fix the documentation, or we could add `allocate`.

**Comment by rossabaker** [comment]
We currently have [this](https://github.com/http4s/http4s/blob/eae95bfa8eb3bda1aafcdc188c200ef20c258723/core/src/main/scala/org/http4s/internal/package.scala#L123-L164) with [these tests](https://github.com/http4s/http4s/blob/eae95bfa8eb3bda1aafcdc188c200ef20c258723/tests/src/test/scala/org/http4s/internal/AllocatedSpec.scala) in http4s.  We had a [simpler version](https://github.com/http4s/http4s/blob/9f2a7d2ed0742d3ce5c0cf1d6d86de021cac24b2/core/src/main/scala/org/http4s/internal/package.scala#L123-L131), but that required a `Concurrent` constraint and hung in failure scenarios.

I'm heads down on http4s-0.20.0 and at a conference next week.  I still intend to add this, but it may about 10 days before I can make time.  

If someone else steps forward sooner, great.  I think testing with cancellation and on something non-IO are the most interesting remaining bits.

**Comment by kubukoz** [comment]
Is this solved by #419?

**Comment by SystemFw** [comment]
it is
